### PR TITLE
IPv6 Hostname

### PR DESCRIPTION
**Pull request by forstermatth**
Adds more support for parsing IPv6 URL's.
Includes changes to the main parse function to remove brackets from IPv6 URL's.

Inspired by:

https://github.com/nodejs/node/blob/a94ffda/lib/url.js#L184
https://github.com/nodejs/node/blob/a94ffda/lib/url.js#L349

### CODE REVIEW

**Comment by forstermatth** [comment]
Still Inconsistent, brackets are missing from host. 

Node: 
```js
Url {
  protocol: 'http:',
  slashes: true,
  auth: null,
  host: '[3ffe:2a00:100:7031::1]',
  port: null,
  hostname: '3ffe:2a00:100:7031::1',
  hash: null,
  search: null,
  query: null,
  pathname: '/foo',
  path: '/foo',
  href: 'http://[3ffe:2a00:100:7031::1]/foo' }
```

url-parse:
```js
URL {
  protocol: 'http:',
  slashes: true,
  hash: '',
  query: '',
  pathname: '/foo',
  auth: '',
  host: '3ffe:2a00:100:7031::1',
  port: '',
  hostname: '3ffe:2a00:100:7031::1',
  password: '',
  username: '',
  href: 'http://[3ffe:2a00:100:7031::1]/foo' }
```

File: index.js [hunk 1]
```diff
@@ -251,7 +280,8 @@ URL.prototype.toString = function toString(stringify) {
     result += '@';
   }
 
-  result += url.hostname;
+  // rough check of IPv6 Hostname (v4 doesn't include colon)
+  result += ~url.hostname.indexOf(':') ? '[' + url.hostname + ']' : url.hostname;
```
Comment (lpinca) [inline index.js]: We can probably replace this line and the next one with just `result += url.host`;

**Comment by lpinca** [comment]
Breaking change but Looks good to me, please rebase against master.

File: index.js [hunk 2]
```diff
@@ -137,6 +148,7 @@ function URL(address, location, parser) {
     if (instruction[4]) {
       url[key] = url[key].toLowerCase();
     }
+
```
Comment (lpinca) [inline index.js]: Please keep the old style, no new line here.

File: index.js [hunk 3]
```diff
@@ -201,12 +216,14 @@ URL.prototype.set = function set(part, value, fn) {
       url.host = url.hostname;
       url[part] = '';
     } else if (value) {
-      url.host = url.hostname +':'+ value;
```
Comment (lpinca) [inline index.js]: We actually use this style, please revert these changes.

File: index.js [hunk 4]
```diff
@@ -219,6 +236,9 @@ URL.prototype.set = function set(part, value, fn) {
       url.hostname = value;
       url.port = '';
     }
+
+    if (isIPv6(url.hostname)) url.hostname = url.hostname.slice(1, -1);
```
Comment (3rd-Eden) [inline index.js]: It could use a comment that ipv6 uses brackets and that we want to remove them from the host name.

File: index.js [hunk 5]
```diff
@@ -198,14 +212,17 @@ URL.prototype.set = function set(part, value, fn) {
     url[part] = value;
 
     if (!required(value, url.protocol)) {
-      url.host = url.hostname;
+      url.host = isIPv6(url.hostname) ? '['+ url.hostname +']' : url.hostname;
       url[part] = '';
     } else if (value) {
-      url.host = url.hostname +':'+ value;
+      url.host = isIPv6(url.hostname) ? '['+ url.hostname +']' +':'+ value : url.hostname +':'+ value;
```
Comment (forstermatth) [inline index.js]: I duplicated efforts here to avoid making a temporary variable, but that was just a guess, a temp variable would clean it up.

```js
var tmpHost = isIPv6(url.hostname) ? '['+ url.hostname +']' : url.hostname;
if (!required(value, url.protocol)) {
  url.host = tmpHost;
  url[part] = '';
} else if (value) {
  url.host = url.hostname +':'+ value;
  url.host = tmpHost +':'+ value;
}

Comment (lpinca) [inline index.js]: Yes, it probably makes sense to use a tmp variable, I would simply call it `host`.

**Comment by lpinca** [comment]
I like the links in your first commit message, but pointing to master is not very useful as the contents will change over time.
We can either remove them or use a specific, abbreviated, commit sha.

**Comment by forstermatth** [comment]
@lpinca Changed!

**Comment by lpinca** [comment]
@forstermatth sorry for being picky but what about:
```
https://github.com/nodejs/node/blob/702997c/lib/url.js#L221
https://github.com/nodejs/node/blob/702997c/lib/url.js#L382
```
so we can directly click or copy/paste without searching the line :)

**Comment by forstermatth** [comment]
@lpinca I found this commit which might serve better node@a94ffdaec57d9999485b11183f8da10aec15ff8d

**Comment by lpinca** [comment]
Awesome, LGTM.

**Comment by lpinca** [comment]
I would cut a new patch release from master then merge this and cut a new major.
Since I can't publish to npm I let @3rd-Eden do this.

**Comment by lpinca** [comment]
@forstermatth good job, thanks.

**Comment by lpinca** [comment]
I've just realized that the old behavior was consistent with the `URL` interface in the browsers. For example in Chrome:
```js
var url = new URL('http://[3ffe:2a00:100:7031::1]:3000/foo')
undefined
url
URL {searchParams: URLSearchParams, href: "http://[3ffe:2a00:100:7031::1]:3000/foo", origin: "http://[3ffe:2a00:100:7031::1]:3000", protocol: "http:", username: ""…}
hash: ""
host: "[3ffe:2a00:100:7031::1]:3000"
hostname: "[3ffe:2a00:100:7031::1]"
href: "http://[3ffe:2a00:100:7031::1]:3000/foo"
origin: "http://[3ffe:2a00:100:7031::1]:3000"
password: ""
pathname: "/foo"
port: "3000"
protocol: "http:"
search: ""
searchParams: URLSearchParams
username: ""
__proto__: URL
```
Now I'm not sure if we should strip the brackets or not 😕.

**Comment by 3rd-Eden** [comment]
@lpinca I think we want to follow the browser as close as possible because that is what we've been doing historically.

**Comment by 3rd-Eden** [comment]
```
a = document.createElement('a')
<a>​</a>​
a.href='http://[3ffe:2a00:100:7031::1]:3000/foo'
"http://[3ffe:2a00:100:7031::1]:3000/foo"
a.host
"[3ffe:2a00:100:7031::1]:3000"
```

**Comment by 3rd-Eden** [comment]
Is there a specific reason why brackets need to be removed from the host? Is there a specification / RFC that we can use as guide?

**Comment by lpinca** [comment]
I also see some advantages in following the browser URL interface:

- less code which leads to better perfomance and smaller size (our main target is the browser anyway)
- no breaking change

I kinda feel bad for @forstermatth who put some effort here :/, but some parts of this pr can be saved.
@3rd-Eden not sure if there is a spec/RFC for this, but we can see what other parsers do like python url parser for example.

**Comment by 3rd-Eden** [comment]
@lpinca the url-parser that ships in python also removes the brackets from the hostname. Seems like there is no general consent here about how to deal with brackets.

**Comment by lpinca** [comment]
Hmm ok, my opinion is that if our main target is the browser it probably makes sense to just follow it (e.g. keep the brackets).

**Comment by lpinca** [comment]
@forstermatth I've cherry-picked some of these changes, see d9dbeedf2dabfb4bb2a16d885b7a38b285754218...e2d2d72c185998c39489604a40500f2c391090bf.
Thanks again.
